src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
+            }
+
+            /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+             * a SC  or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+            }
+
+            /*  Point of no return */
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+            endpoint_ptr_set_epQueue_head_np(ep_ptr, TCB_REF(dest->tcbEPNext));
+            if (unlikely(dest->tcbEPNext)) {
+                dest->tcbEPNext->tcbEPPrev = NULL;
+            } else {
+                endpoint_ptr_mset_epQueue_tail_state(ep_ptr, 0, EPState_Idle);
+            }
+
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            if (!dest->tcbSchedContext) {
+                schedContext_donate(sc, dest);
+            }
+            assert(dest->tcbSchedContext);
+
+            /* Enqueue signalled but continue executing signaller */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+            tcbSchedAppend(dest);
+            restore_user_context();
+            UNREACHABLE();
+        } else {
+            ntfn_set_active(ntfnPtr, badge);
+            restore_user_context();
+            UNREACHABLE();
+        }
+        break;
+    }
+    case NtfnState_Waiting: {
+
+        /* Signal to higher prio thread is NOT fastpathed*/
+        if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+            slowpath(SysSend);
+        }
+
+        /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+         * a SC or if one can be donated from the notification. If neither is true, go to
+         * slowpath */
+        if (!dest->tcbSchedContext) {
+            sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+            if (sc == NULL || sc->scTcb != NULL) {
+                slowpath(SysSend);
+            }
+        }
+
+        /*  Point of no return */
+
+        tcb_queue_t ntfn_queue;
+        ntfn_queue.head = (tcb_t *)notification_ptr_get_ntfnQueue_head(ntfnPtr);
+        ntfn_queue.end = (tcb_t *)notification_ptr_get_ntfnQueue_tail(ntfnPtr);
+
+        ntfn_queue = tcbEPDequeue(dest, ntfn_queue);
+
+        notification_ptr_set_ntfnQueue_head(ntfnPtr, (word_t)ntfn_queue.head);
+        notification_ptr_set_ntfnQueue_tail(ntfnPtr, (word_t)ntfn_queue.end);
+
+        if (!ntfn_queue.head) {
+            notification_ptr_set_state(ntfnPtr, NtfnState_Idle);
+        }
```

Comment:
Is this intentionally different from the other fastpath queue modification code below? Notification objects have the same bitfield structure as endpoints for the `ntfnQueue_head`, `ntfnQueue_tail` and `state` fields, so the same fastpath handling done for endpoints should also be possible (just with `notification_*` function variants):

```
    /* Dequeue the destination. */
    endpoint_ptr_set_epQueue_head_np(ep_ptr, TCB_REF(dest->tcbEPNext));
    if (unlikely(dest->tcbEPNext)) {
        dest->tcbEPNext->tcbEPPrev = NULL;
    } else {
        endpoint_ptr_mset_epQueue_tail_state(ep_ptr, 0, EPState_Idle);
    }

```

Reply:
I might have argued a bit too strongly for always using the bitfield accessors before. If we have the same pattern in the EP fastpath, we should be able to deal with it in the verification.

Reply:
Let's ignore this comment for this PR.